x/accounts/vesting/vesting.go

```diff
@@ -0,0 +1,490 @@
+			err = bva.DelegatedFree.Set(ctx, newDelFree.Denom, newDelFree.Amount)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// TrackUndelegation tracks an undelegation amount by setting the necessary
+// values by which delegated vesting and delegated vesting need to decrease and
+// by which amount the base coins need to increase.
+//
+// NOTE: The undelegation (bond refund) amount may exceed the delegated
+// vesting (bond) amount due to the way undelegation truncates the bond refund,
+// which can increase the validator's exchange rate (tokens/shares) slightly if
+// the undelegated tokens are non-integral.
+//
+// CONTRACT: The account's coins and undelegation coins must be sorted.
+func (bva *BaseVesting) TrackUndelegation(ctx context.Context, amount sdk.Coins) error {
+	for _, coin := range amount {
+		// return error if the undelegation amount is zero
+		if coin.Amount.IsZero() {
+			return sdkerrors.ErrInvalidCoins.Wrap("undelegation attempt with zero coins")
+		}
+		delFreeAmt, err := bva.DelegatedFree.Get(ctx, coin.Denom)
+		if err != nil {
+			return err
+		}
+		delVestingAmt, err := bva.DelegatedVesting.Get(ctx, coin.Denom)
+		if err != nil {
+			return err
+		}
+
+		// compute x and y per the specification, where:
+		// X := min(DF, D)
+		// Y := min(DV, D - X)
+		x := math.MinInt(delFreeAmt, coin.Amount)
+		y := math.MinInt(delVestingAmt, coin.Amount.Sub(x))
+
+		delVestingCoin := sdk.NewCoin(coin.Denom, delVestingAmt)
+		delFreeCoin := sdk.NewCoin(coin.Denom, delFreeAmt)
+		if !x.IsZero() {
+			xCoin := sdk.NewCoin(coin.Denom, x)
+			newDelFree := delFreeCoin.Sub(xCoin)
+			err = bva.DelegatedVesting.Set(ctx, newDelFree.Denom, newDelFree.Amount)
+			if err != nil {
+				return err
+			}
+		}
+
+		if !y.IsZero() {
+			yCoin := sdk.NewCoin(coin.Denom, y)
+			newDelVesting := delVestingCoin.Sub(yCoin)
+			err = bva.DelegatedVesting.Set(ctx, newDelVesting.Denom, newDelVesting.Amount)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// ExecuteMessages handle the execution of codectypes Any messages
+// and update the vesting account DelegatedFree and DelegatedVesting
+// when delegate or undelegate is trigger. And check for locked coins
+// when performing a send message.
+func (bva *BaseVesting) ExecuteMessages(
+	ctx context.Context, msg *account_abstractionv1.MsgExecute, getVestingFunc getVestingFunc,
+) (
+	*account_abstractionv1.MsgExecuteResponse, error,
+) {
+	// we always want to ensure that this is called by the x/accounts module, it's the only trusted entrypoint.
+	// if we do not do this check then someone could call this method directly and bypass the authentication.
+	if !accountstd.SenderIsAccountsModule(ctx) {
+		return nil, fmt.Errorf("sender is not the x/accounts module")
+	}
+	hs := bva.headerService.GetHeaderInfo(ctx)
+
+	for _, m := range msg.ExecutionMessages {
+		typeUrl := m.TypeUrl
+		switch typeUrl {
+		case "/cosmos.staking.v1beta1.MsgDelegate":
```

Comment:
this is not safe as a msg can come as "google.com/cosmos...", there's a specific way to parse this typeURL